### PR TITLE
SQS backed diff/actuate

### PR DESCRIPTION
**Pull request by robfletcher**

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -4,6 +4,7 @@ coroutinesVersion=1.1.1
 hikariVersion=2.5.1
 kotlinVersion=1.3.21
 liquibaseVersion=3.6.3
+localstackVersion=1.10.6
```
Comment (robfletcher) [inline gradle.properties]: Travis is not having this so I'll try to find an alternative.

File: keel-sqs/src/main/kotlin/com/netflix/spinnaker/keel/sqs/SqsResourceCheckQueue.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import com.amazonaws.services.sqs.AmazonSQS
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.config.SqsProperties
+import com.netflix.spinnaker.keel.annealing.ResourceCheckQueue
+import com.netflix.spinnaker.keel.api.ApiVersion
+import com.netflix.spinnaker.keel.api.ResourceName
+import javax.annotation.PostConstruct
+
+internal class SqsResourceCheckQueue(
+  private val sqs: AmazonSQS,
```
Comment (robfletcher) [inline keel-sqs/src/main/kotlin/com/netflix/spinnaker/keel/sqs/SqsResourceCheckQueue.kt]: I'm not actually creating this bean anywhere. Trying to figure that out.

File: keel-sqs/src/main/kotlin/com/netflix/spinnaker/config/SqsProperties.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.netflix.spinnaker.config
+
+import org.springframework.boot.context.properties.ConfigurationProperties
+
+@ConfigurationProperties("sqs")
+class SqsProperties {
+  var queueARN: String = ""
```
Comment (emjburns) [inline keel-sqs/src/main/kotlin/com/netflix/spinnaker/config/SqsProperties.kt]: Do we also need to specify the topic ARN?

Comment (emjburns) [reply keel-sqs/src/main/kotlin/com/netflix/spinnaker/config/SqsProperties.kt]: Or, are you just creating a queue and pushing messages and reading from the queue?

File: keel-sqs/src/test/kotlin/com/netflix/spinnaker/keel/sqs/LocalStackSqsConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.amazonaws.services.sqs.AmazonSQS
+import com.amazonaws.services.sqs.AmazonSQSClient
+import com.netflix.spinnaker.keel.serialization.configuredObjectMapper
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.testcontainers.containers.localstack.LocalStackContainer
+
+@Configuration
+class LocalStackSqsConfiguration {
```
Comment (robfletcher) [inline keel-sqs/src/test/kotlin/com/netflix/spinnaker/keel/sqs/LocalStackSqsConfiguration.kt]: Just split this out into its own file

File: keel-sqs/src/test/kotlin/com/netflix/spinnaker/keel/sqs/SqsResourceCheckListenerTests.kt [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import com.nhaarman.mockitokotlin2.doThrow
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.never
+import com.nhaarman.mockitokotlin2.reset
+import com.nhaarman.mockitokotlin2.stub
+import com.nhaarman.mockitokotlin2.verify
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+
+internal class SqsResourceCheckListenerTests : JUnit5Minutests {
```
Comment (robfletcher) [inline keel-sqs/src/test/kotlin/com/netflix/spinnaker/keel/sqs/SqsResourceCheckListenerTests.kt]: Faster running unit tests that don't deal with any concurrency / coroutine management and just do a single poll on the SQS queue.

File: keel-sqs/src/main/kotlin/com/netflix/spinnaker/config/SqsConfiguration.kt [hunk 6]
```diff
@@ -53,5 +53,7 @@ class SqsConfiguration {
     objectMapper: ObjectMapper,
     actuator: ResourceActuator
   ): Any =
-    SqsResourceCheckListener(sqsClient, queueARN, objectMapper, actuator)
+    sqsClient.createQueue(queueARN.name).run {
```
Comment (robfletcher) [inline keel-sqs/src/main/kotlin/com/netflix/spinnaker/config/SqsConfiguration.kt]: Feel like there might be a better place to do the queue creation than here but probably ok for now.

File: keel-sqs/src/test/kotlin/com/netflix/spinnaker/keel/sqs/SqsIntegrationTests.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+@SpringBootTest(
+  classes = [LocalStackSqsConfiguration::class, SqsConfiguration::class],
+  properties = [
+    "sqs.enabled=true",
+    "sqs.account-name=test",
+    "sqs.queue-arn=arn:aws:sqs:ap-south-1:530822479253:keel-test-resource_check"
+  ],
+  webEnvironment = NONE
+)
+internal class SqsIntegrationTests {
```
Comment (robfletcher) [inline keel-sqs/src/test/kotlin/com/netflix/spinnaker/keel/sqs/SqsIntegrationTests.kt]: Integration test that uses localstack to stand up an SQS API and make sure the queue read / write work in congress.

Thoughts on whether this has value? It's pretty slow due to localstack pulling down a fucking giant Docker image.

Comment (emjburns) [reply keel-sqs/src/test/kotlin/com/netflix/spinnaker/keel/sqs/SqsIntegrationTests.kt]: I don't feel like it has a ton of value. What cases would this docker image version of SQS catch that we wouldn't catch another way? Seems like increased build time is a win regardless.

**Pull request merged**